### PR TITLE
Avoid registering as keywords the #... in Primitive

### PR DESCRIPTION
**Pull request by proux01**
This was discussed in #11117 . I don't know if we want to do that, but at least it seems feasible.

### CODE REVIEW

**Comment by maximedenes** [comment]
This `#` sign is a signal that the name is built in. Can you explain here what the motivation for removing it?

**Comment by proux01** [comment]
I'm personally fine with it (although I wonder whether it's not somewhat redundant with the `Primitive` keyword?) but that seem to have an undesirable side effect according to this comment https://github.com/coq/coq/pull/11117#discussion_r347032445

**Comment by maximedenes** [comment]
Sorry, but could you (or somebody else) post a summary? The discussion is very long there. I see that primitive names with `#` are keywords, but I don't know what's problematic with it.

**Comment by herbelin** [comment]
>  I see that primitive names with `#` are keywords, but I don't know what's problematic with it.

I don't think this is anything really problematic with them. The question was raised at the beginning in relation with the objectie to list in the manual the reserved keywords of Gallina. Then, the question was raised of explainng the int63/float64 `#` keywords. Then, it was realized that they did not need to be keywords since they already occur in a well-defined context introduced by the command `Primitive`.

So, it is not a problem that they are keywords, in the sense that they are so specific that the risk of collision with a development needing exactly the same expression of a form like `#int63_lsr` seems pretty weak. On the other side, they don't need to be keywords either so, the description of the Gallina grammar would benefit of a simpler presentation by not declaring them as keywords.

**Comment by maximedenes** [comment]
I have no particular opinion on the distinction about keywords / non-keywords, or the way these built-ins are presented in the manual. On the other hand, I think we should keep the leading `#` as a way to distinguish them from regular identifiers. Thus, I would lean towards closing this PR.

**Comment by herbelin** [comment]
> On the other hand, I think we should keep the leading # as a way to distinguish them from regular identifiers.

If you want to have a `#` in front of them in the syntax, one should be able to do it with a look-ahead. Concretely, that should be something like the following, and that should prevent declaring keywords:
```
| IDENT "Primitive"; id = identref; typopt = OPT [ ":"; typ = lconstr -> { typ } ]; 
":="; test_hash_ident; "#"; r = register_token -> ...
```
with
```
let test_hash_ident =                                                                                         
  let open Pcoq.Lookahead in                                                                                        
  to_entry "test_hash_ident" begin                                                                            
    lk_kw "#" >> lk_ident >> check_no_space                                                                        
  end    
```
Not tested though, but since the look-ahead is working on the lexemes, it should be able to see the `IDENT` which `register_token` will parse.

**Comment by Zimmi48** [comment]
This looks to me like an overly complex solution. If we want these not to be keywords, we could just not list them in the parser at all, parse an ident (after the leading `#`) and raise an error at interpretation time if the primitive identifier is not among the known primitives, couldn't we?

**Comment by proux01** [comment]
So here is a solution that keeps the `#` and should not clutter the doc.

**Comment by proux01** [comment]
> This looks to me like an overly complex solution. If we want these not to be keywords, we could just not list them in the parser at all, parse an ident (after the leading `#`) and raise an error at interpretation time if the primitive identifier is not among the known primitives, couldn't we?

Sure, the lookahead is just required to avoid accepting something with extra spaces like `# int63_type` instead of `#int63_type`.

File: vernac/g_vernac.mlg [hunk 1]
```diff
@@ -226,63 +231,12 @@ GRAMMAR EXTEND Gram
...
-       | "#float64_next_up" -> { CPrimitives.Float64next_up }
-       | "#float64_next_down" -> { CPrimitives.Float64next_down }
-       ] ]
-    ;
-
+  register_token:
+    [ [ test_hash_ident; "#"; r = IDENT ->
+          { try CPrimitives.(OT_type (prim_type_of_string r))
+            with Not_found -> try CPrimitives.(OT_op (parse r))
+            with Not_found -> user_err ~loc Pp.(str "register_token expected.") } ] ]
```
Comment (maximedenes) [inline vernac/g_vernac.mlg:238]: The PR looks very good, thanks. Maybe we can just seize the opportunity of improving the error message here? Something like `Built-in #foo does not exist.`, maybe?

File: vernac/g_vernac.mlg [hunk 2]
```diff
@@ -226,63 +231,12 @@ GRAMMAR EXTEND Gram
...
-       | "#float64_frshiftexp" -> { CPrimitives.Float64frshiftexp }
-       | "#float64_ldshiftexp" -> { CPrimitives.Float64ldshiftexp }
-       | "#float64_next_up" -> { CPrimitives.Float64next_up }
-       | "#float64_next_down" -> { CPrimitives.Float64next_down }
-       ] ]
-    ;
-
+  register_token:
+    [ [ test_hash_ident; "#"; r = IDENT ->
+          { try CPrimitives.(OT_type (prim_type_of_string r))
```
Comment (SkySkimmer) [inline vernac/g_vernac.mlg:236]: I don't think we should do this lookup at parsing time.

Comment (proux01) [reply vernac/g_vernac.mlg:236]: Do you have a rationale? It seems to me that these built-ins are hardcoded by nature.

Comment (SkySkimmer) [reply vernac/g_vernac.mlg:236]: I don't like having code in the parser, a couple trywith blocks is complicated enough for me to notice

File: vernac/g_vernac.mlg (hunk 1 above)
Comment (proux01) [reply vernac/g_vernac.mlg]: Done, thanks!

File: vernac/g_vernac.mlg (hunk 2 above)
Comment (proux01) [reply vernac/g_vernac.mlg]: We can hide it into CPrimitives if you prefer. Do you agree with this new version?

Comment (SkySkimmer) [reply vernac/g_vernac.mlg]: It's ok

Comment (proux01) [reply vernac/g_vernac.mlg]: Thanks!

**Comment by Zimmi48** [comment]
CI failure is unrelated.

**Review by maximedenes — APPROVED** [review]

**Pull request merged**